src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -802,58 +875,72 @@ private static bool TryParseExactX(ReadOnlySpan<char> guidString, ref GuidResult
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static byte DecodeByte(char ch1, char ch2, ref int invalidIfNegative)
+        private static byte DecodeByte<TChar>(TChar ch1, TChar ch2, ref int invalidIfNegative) where TChar : unmanaged, IUtfChar<TChar>
         {
             ReadOnlySpan<byte> lookup = HexConverter.CharToHexLookup;
             Debug.Assert(lookup.Length == 256);
+            Debug.Assert(ch1 is byte or char);
 
-            int upper = (sbyte)lookup[(byte)ch1];
-            int lower = (sbyte)lookup[(byte)ch2];
-            int result = (upper << 4) | lower;
+            if (ch1 is char c1 && ch2 is char c2)
+            {
+                int upper = (sbyte)lookup[(byte)c1];
+                int lower = (sbyte)lookup[(byte)c2];
```

Comment:
This is inefficient and potentially boxing.
DO NOT use type pattern matching on generic typed arguments. Instead, leverage the `IUtfChar` interface to convert the value to integer.
It should also be unified with `byte` path.

Reply:
Is it even possible to make it box if `TChar` is `char`?

Reply:
For legacy JIT, yes. It may also box for mono.

Using IUtfChar method is the consistent pattern for UTF8/16 unification.

Reply:
Mostly .NET Framework. Yes ryujit is able to handle this, but I'm not sure about Mono, since there would be `box` instructions in generated IL.